src/main/scala/com/microsoft/azure/cosmosdb/spark/ClientCacheEntry.scala

```diff
@@ -0,0 +1,84 @@
+  * copies or substantial portions of the Software.
+  *
+  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+  * SOFTWARE.
+  */
+package com.microsoft.azure.cosmosdb.spark
+
+import com.microsoft.azure.documentdb._
+import com.microsoft.azure.documentdb.bulkexecutor.DocumentBulkExecutor
+
+/**
+  * Case class used as an envelope of all the data being cached in teh Executors for usage
+  * in CosmosDBConnection within.
+  *
+  * @param docClient                An instance of the DocumentClient used by the CosmosDBConnection
+  * @param bulkExecutor             An instance of the BulkExecutor instance used by the CosmosDBConnection (optional).
+  *                                 Will be lazily initialized only when needed.
+  * @param containerMetadata        Container metadata for the target container. Only immutable information is stored
+  *                                 here - so does not to be refreshed or updated on partition splits etc. (optional).
+  *                                 Will be lazily initialized only when needed.
+  * @param databaseMetadata         Database metadata for the database. Only immutable information is stored
+  *                                 here - so does not to be refreshed or updated on partition splits etc. (optional)
+  *                                 Will be lazily initialized only when needed.
+  * @param maxAvailableThroughput   The effectively available provisioned throughput. It
+  *                                 is calculated by determining whether the throughput is provisioned at the container
+  *                                 or database level and the configured max. throughput that may be used for bulk
+  *                                 operations. This information is mutable - and is going to be refreshed regularly to
+  *                                 be bale to react and if possible increase throughput used by bulk executor if
+  *                                 additional throughput is getting provisioned.
+  */
+private[spark] case class ClientCacheEntry(
+                            docClient: DocumentClient,
+                            bulkExecutor: Option[DocumentBulkExecutor],
+                            containerMetadata: Option[ContainerMetadata],
+                            databaseMetadata: Option[DatabaseMetadata],
+                            maxAvailableThroughput: Option[Int]) {
+
+  def getLogMessage : String = {
+    val clientMsg = s"DocumentClient#${docClient.hashCode()}"
```

Comment:
why do we need the hashcode to be part of the log message?

Reply:
This pattern has been super useful when debugging issues with the connection pooling (when unintentionally multiple instances hve been instantiated etc. when incorrectly using case classes etc.) So I am basically using it as a way to differentiate multiple instances of DocumentClient etc. - is there any more elegant way of doing this in java or any concern with the approach?